### PR TITLE
feat: add the whatsapp clone app to projects

### PR DESCRIPTION
**Pull request by vanGalilea**
I've been working on this project for a while now and would like to get your feedback regarding things that couldve been done better or any weird quirks.

Tnx in advance!

### CODE REVIEW

File: projects/WhatsAppClone/src/components/channel-list/ChannelListHeader.tsx [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+      style={{
+        backgroundColor: colors.dark.secondary,
+        ...flex.directionRowItemsCenter,
+      }}>
+      <StatusBar backgroundColor={colors.dark.secondary} />
+      <PeekabooView isEnabled={isInChannelelectionMode}>
+        <View style={flex.directionRowItemsCenterContentSpaceBetween1}>
+          <View style={flex.directionRowItemsCenter}>
+            <IconButton
+              // onPress={clearSelectedMessageIdsEditing}
```
Comment (vanGalilea) [inline projects/WhatsAppClone/src/components/channel-list/ChannelListHeader.tsx:32]: fix this

File: projects/WhatsAppClone/src/components/channel-list/ChannelListHeader.tsx [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+              pathFill={colors.dark.text}
+            />
+            <Text
+              numberOfLines={1}
+              style={{
+                color: colors.dark.text,
+                fontWeight: 'bold',
+                fontSize: sizes.l,
+              }}>
+              2{/*{selectedMessageIdsEditing.length}*/}
```
Comment (vanGalilea) [inline projects/WhatsAppClone/src/components/channel-list/ChannelListHeader.tsx:43]: user var here

File: projects/WhatsAppClone/src/components/channel-list/ChannelListHeader.tsx [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+            </Text>
+          </View>
+
+          <View style={flex.directionRowItemsCenter}>
```
Comment (vanGalilea) [inline projects/WhatsAppClone/src/components/channel-list/ChannelListHeader.tsx:47]: remove or leave without //

File: projects/WhatsAppClone/src/screens/WallpaperTypeDetails.tsx [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+  'WallpaperTypeDetails'
+>
+
+export type Props = {
+  navigation: CustomWallPaperScreenNavigationProp
+  route: CustomWallPaperRouteProp
+}
+
+const GRID_ITEM_WIDTH = '32.7%'
+const SOLID_COLORS = [
```
Comment (vanGalilea) [inline projects/WhatsAppClone/src/screens/WallpaperTypeDetails.tsx:28]: move to external files and tidy up styles

File: projects/WhatsAppClone/src/screens/WallpaperTypesOverview.tsx [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+  },
+}: Props) => {
+  const {clearPreferences} = useChannelPreferences(channelId)
+
+  const handleRemoveCustomBackground = () => {
+    clearPreferences()
+    navigate('CustomWallpaper', {channelId})
+  }
+  return (
+    <>
```
Comment (vanGalilea) [inline projects/WhatsAppClone/src/screens/WallpaperTypesOverview.tsx:59]: tidy up styles and move to StyleSheet

File: projects/WhatsAppClone/src/theme.ts [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+    highlighted: '#1a2427',
+    transparentPrimary: 'rgba(255,255,255,0.5)',
+    primaryLightTransparent: 'rgba(3,91,73,0.6)',
+    border: '#323b40',
+    danger: '#D3514CFF',
+    active: '#52b9e3',
+  },
+}
+
+export const theme: DeepPartial<Theme> = {
```
Comment (vanGalilea) [inline projects/WhatsAppClone/src/theme.ts:24]: reduce duplic.

File: projects/WhatsAppClone/src/client.ts [hunk 7]
```diff
@@ -0,0 +1,11 @@
+import {StreamChat} from 'stream-chat'
+import {STREAM_API_KEY, STREAM_USER_TOKEN, STREAM_USER_ID} from '@env'
+import {ChatContextValue} from 'stream-chat-react-native'
+
+export const chatClient = StreamChat.getInstance(
+  STREAM_API_KEY,
+) as unknown as ChatContextValue['client']
```
Comment (madsroskar) [inline projects/WhatsAppClone/src/client.ts:7]: Look into why this is needed

File: projects/WhatsAppClone/src/components/PeekabooView.tsx [hunk 8]
```diff
@@ -0,0 +1,10 @@
+import React, {Fragment} from 'react'
+import {ViewProps} from 'react-native'
+
+interface Props extends ViewProps {
+  isEnabled: boolean | undefined
+}
+export default ({isEnabled, ...props}: Props) => {
+  if (!isEnabled) return null
+  return <Fragment {...props} />
```
Comment (madsroskar) [inline projects/WhatsAppClone/src/components/PeekabooView.tsx:9]: Fragments aren't usually rendered - should children be rendered instead?

File: projects/WhatsAppClone/src/components/PressMe.tsx [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+import {TapGestureHandler} from 'react-native-gesture-handler'
+import useRippleEffect from '../hooks/useRippleEffect'
+
+type Props = ViewProps & {onPress: PressableProps['onPress']}
+
+export default ({children, onPress, style = {}, ...props}: Props) => {
+  const {ref, tapGestureEvent, rStyle} = useRippleEffect({onPress})
+
+  return (
+    <View ref={ref} {...props} style={{zIndex: 5}}>
```
Comment (madsroskar) [inline projects/WhatsAppClone/src/components/PressMe.tsx:13]: Why is it 5? :D

File: projects/WhatsAppClone/src/components/PressMe.tsx [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+      <TapGestureHandler onGestureEvent={tapGestureEvent}>
+        <Animated.View style={{overflow: 'hidden', ...(style as object)}}>
```
Comment (madsroskar) [inline projects/WhatsAppClone/src/components/PressMe.tsx:15]:
```suggestion
        <Animated.View style={[{overflow: 'hidden'}, style]}>
```

Maybe this avoids the assertion?

File: projects/WhatsAppClone/src/components/SuperAvatar.tsx [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+  outerContainer: {flex: 0},
+  checkWrap: {
+    padding: 2,
+    borderRadius: 16,
+    borderWidth: 2,
+    borderColor: colors.dark.highlighted,
+    backgroundColor: colors.dark.primaryLight,
+    position: 'absolute',
+    bottom: -4,
+    right: -4,
```
Comment (madsroskar) [inline projects/WhatsAppClone/src/components/SuperAvatar.tsx:40]: 😳

File: projects/WhatsAppClone/src/components/channel-list/ChannelPreview.tsx [hunk 12]
```diff
@@ -0,0 +1,195 @@
+import {Pressable, StyleSheet, Text, View} from 'react-native'
+import React, {useContext, useMemo, useState} from 'react'
+import {useChannelPreviewDisplayName} from 'stream-chat-react-native-core/src/components/ChannelPreview/hooks/useChannelPreviewDisplayName'
```
Comment (madsroskar) [inline projects/WhatsAppClone/src/components/channel-list/ChannelPreview.tsx:3]: These autoimports look a bit wonky

File: projects/WhatsAppClone/src/components/channel/ChannelHeader.tsx [hunk 13]
```diff
@@ -0,0 +1,206 @@
+import {Alert, SafeAreaView, StyleSheet, Text, View} from 'react-native'
+import {colors} from '../../theme'
+import React, {useContext, useMemo} from 'react'
+import {AppContext} from '../../App'
+import {flex, sizes} from '../../global'
+import IconButton from '../IconButton'
+import SuperAvatar from '../SuperAvatar'
+import {useChannelPreviewDisplayName} from 'stream-chat-react-native-core/src/components/ChannelPreview/hooks/useChannelPreviewDisplayName'
```
Comment (madsroskar) [inline projects/WhatsAppClone/src/components/channel/ChannelHeader.tsx:8]:
```suggestion
import {useChannelPreviewDisplayName} from 'stream-chat-react-native-core/src/components/ChannelPreview/hooks/useChannelPreviewDisplayName'
```

😳

File: projects/WhatsAppClone/src/components/channel/MessageInput.tsx [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+          </PeekabooView>
+          <PeekabooView isEnabled={!recordingActive}>
+            <IconButton
+              onPress={() => null}
+              iconName={'Smiley'}
+              pathFill={colors.dark.secondaryLight}
+            />
+            <AutoCompleteInput
+              setInputBoxRef={
+                messageInputRef as unknown as LocalMessageInputContext['setInputBoxRef']
```
Comment (madsroskar) [inline projects/WhatsAppClone/src/components/channel/MessageInput.tsx:145]: Why is this needed ? 🤔

File: projects/WhatsAppClone/src/components/channel/MessageInputCTA.tsx [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+} from 'stream-chat-react-native'
+import {StyleSheet, View} from 'react-native'
+import {globalStyles, sizes} from '../../global'
+import {colors} from '../../theme'
+import IconButton from '../IconButton'
+import AudioRecorderPlayer from 'react-native-audio-recorder-player'
+import moment from 'moment'
+import {MessageResponse} from 'stream-chat'
+import {set} from 'lodash'
+import {Message} from 'stream-chat/src/types'
```
Comment (madsroskar) [inline projects/WhatsAppClone/src/components/channel/MessageInputCTA.tsx:16]: should this be a type import? Also, autoimported?

File: projects/WhatsAppClone/src/components/channel/MessageInputCTA.tsx [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+      mentioned_users: [],
+      id: `random-id-${new Date().toTimeString()}`,
+      status: 'sending',
+      type: 'regular',
+      user: client.user,
+    }
+    updateMessage(message)
+
+    const res = await channel.sendFile(uri, 'test.mp4', 'audio/mp4')
+    /* eslint-disable @typescript-eslint/no-unused-vars */
```
Comment (madsroskar) [inline projects/WhatsAppClone/src/components/channel/MessageInputCTA.tsx:65]: are any actually unused? 🤔

File: projects/WhatsAppClone/src/components/channel/MessageInputCTA.tsx [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+    const {
+      created_at,
+      html,
+      type,
+      status,
+      user,
+      ...messageWithoutReservedFields
+    } = message
+    /* eslint-enable @typescript-eslint/no-unused-vars  */
```
Comment (madsroskar) [inline projects/WhatsAppClone/src/components/channel/MessageInputCTA.tsx:74]: and here

File: projects/WhatsAppClone/src/components/channel/MessageText.tsx [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+import {flex, sizes} from '../../global'
+import Star from '../../icons/Star'
+import {colors} from '../../theme'
+import {MessageStatus, useMessageContext} from 'stream-chat-react-native'
+import React, {useMemo} from 'react'
+import {MessageTextProps} from 'stream-chat-react-native-core/lib/typescript/components/Message/MessageSimple/MessageTextContainer'
+import PeekabooView from '../PeekabooView'
+import {StreamChatGenerics} from '../../types'
+import {RenderTextParams} from 'stream-chat-react-native-core/lib/typescript/components/Message/MessageSimple/utils/renderText'
+import {isEmpty} from 'lodash'
```
Comment (madsroskar) [inline projects/WhatsAppClone/src/components/channel/MessageText.tsx:13]: Autoimports

File: projects/WhatsAppClone/src/components/channel/VoiceMessageAttachment.tsx [hunk 19]
```diff
@@ -0,0 +1,187 @@
+import React, {useMemo, useState} from 'react'
+import {StyleSheet, Text, View} from 'react-native'
+import {
+  Avatar,
+  MessageStatus,
+  MessageType,
+  useMessageContext,
+} from 'stream-chat-react-native'
+import {CardProps} from 'stream-chat-react-native-core/lib/typescript/components/Attachment/Card'
```
Comment (madsroskar) [inline projects/WhatsAppClone/src/components/channel/VoiceMessageAttachment.tsx:9]: Autoimports

File: projects/WhatsAppClone/src/components/channel/VoiceMessageAttachment.tsx [hunk 20]
```diff
@@ -0,0 +1,187 @@
...
+  const onPausePlay = async () => {
+    await pausePlayer()
+  }
+
+  const onStopPlay = async () => {
+    setPaused(false)
+    setCurrentPositionInSeconds(0)
+  }
+
+  // @ts-ignore
```
Comment (madsroskar) [inline projects/WhatsAppClone/src/components/channel/VoiceMessageAttachment.tsx:68]: 😳

File: projects/WhatsAppClone/src/components/channel/VoiceMessageAttachment.tsx [hunk 21]
```diff
@@ -0,0 +1,187 @@
...
+  }
+
+  const onPausePlay = async () => {
+    await pausePlayer()
+  }
+
+  const onStopPlay = async () => {
+    setPaused(false)
+    setCurrentPositionInSeconds(0)
+  }
```
Comment (madsroskar) [inline projects/WhatsAppClone/src/components/channel/VoiceMessageAttachment.tsx:66]: Doesn't need to be async?

File: projects/WhatsAppClone/src/hooks/useRippleEffect.ts [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+        centerX.value = tapEvent.x
+        centerY.value = tapEvent.y
+
+        rippleOpacity.value = 1
+        scale.value = 0
+        scale.value = withTiming(1, {duration: 1000})
+      },
+      onActive: () => {
+        if (onPress) {
+          // @ts-ignore
```
Comment (madsroskar) [inline projects/WhatsAppClone/src/hooks/useRippleEffect.ts:39]: 😳

File: projects/WhatsAppClone/src/screens/Channel.tsx [hunk 23]
```diff
@@ -0,0 +1,37 @@
...
+import MessageInput from '../components/channel/MessageInput'
+import ChannelHeader from '../components/channel/ChannelHeader'
+import ChannelBackgroundWrapper from '../utils/ChannelBackgroundWrapper'
+import {colors} from '../theme'
+import {ActivityIndicator, View} from 'react-native'
+import {flex} from '../global'
+
+export default () => {
+  const {
+    // @ts-ignore
```
Comment (madsroskar) [inline projects/WhatsAppClone/src/screens/Channel.tsx:13]: 😳

File: projects/WhatsAppClone/src/screens/ChannelList.tsx [hunk 24]
```diff
@@ -0,0 +1,108 @@
+import React, {useState} from 'react'
+import {StyleSheet, Text, useWindowDimensions, View} from 'react-native'
+import {SceneMap, TabBar, TabView} from 'react-native-tab-view'
+import {colors} from '../theme'
+import {SceneRendererProps} from 'react-native-tab-view/src/types'
+import {
+  NavigationState,
+  Scene,
+} from 'react-native-tab-view/lib/typescript/types'
```
Comment (madsroskar) [inline projects/WhatsAppClone/src/screens/ChannelList.tsx:9]: Autoimports

File: projects/WhatsAppClone/src/screens/ChannelList.tsx [hunk 25]
```diff
@@ -0,0 +1,108 @@
...
+
+export default () => {
+  const layout = useWindowDimensions()
+  const [index, setIndex] = useState(1)
+  const [routes] = useState([
+    {key: 'camera'},
+    {key: 'chats', title: 'CHATS'},
+    {key: 'status', title: 'STATUS'},
+    {key: 'calls', title: 'CALLS'},
+  ])
```
Comment (madsroskar) [inline projects/WhatsAppClone/src/screens/ChannelList.tsx:35]: useState seems a bit odd here, no? 🤔

File: projects/WhatsAppClone/src/screens/ChannelList.tsx [hunk 26]
```diff
@@ -0,0 +1,108 @@
...
+    focused,
+    route,
+  }: Scene<any> & {
+    focused: boolean
+    color: string
+  }) => {
+    const color = focused
+      ? colors.dark.primaryLight
+      : colors.dark.secondaryLight
+    const itemWidth = layout.width / 4.2
```
Comment (madsroskar) [inline projects/WhatsAppClone/src/screens/ChannelList.tsx:47]: 4.2 😳

File: projects/WhatsAppClone/src/screens/ImagePreview.tsx [hunk 27]
```diff
@@ -0,0 +1,99 @@
...
+  View,
+} from 'react-native'
+import {get, isEmpty} from 'lodash'
+import React, {useMemo} from 'react'
+import {StackNavigationProp} from '@react-navigation/stack'
+import {flex, sizes} from '../global'
+import IconButton from '../components/IconButton'
+import {colors} from '../theme'
+import {useMessageInputContext} from 'stream-chat-react-native'
+import {AutoCompleteInput} from 'stream-chat-react-native-core/src/components/AutoCompleteInput/AutoCompleteInput'
```
Comment (madsroskar) [inline projects/WhatsAppClone/src/screens/ImagePreview.tsx:15]: Autoimport

File: projects/WhatsAppClone/src/screens/WallpaperTypeDetails.tsx [hunk 28]
```diff
@@ -0,0 +1,99 @@
...
+  StackNavigatorParamList,
+  'WallpaperTypeDetails'
+>
+
+export type Props = {
+  navigation: CustomWallPaperScreenNavigationProp
+  route: CustomWallPaperRouteProp
+}
+
+const GRID_ITEM_WIDTH = '32.7%'
```
Comment (madsroskar) [inline projects/WhatsAppClone/src/screens/WallpaperTypeDetails.tsx:28]: LOVE IT

File: projects/WhatsAppClone/src/utils/WhatsAppChannelWrapper.tsx [hunk 29]
```diff
@@ -0,0 +1,101 @@
...
+} from 'stream-chat-react-native'
+import {myMessageTheme} from '../theme'
+import RenderNothing from '../components/RenderNothing'
+import Reply from '../components/channel/Reply'
+import {isEmpty} from 'lodash'
+import MessageContent from '../components/channel/MessageContent'
+import MessageText from '../components/channel/MessageText'
+import VoiceMessageAttachment from '../components/channel/VoiceMessageAttachment'
+import {StreamChatGenerics} from '../types'
+import {ChannelProps} from 'stream-chat-react-native-core/src/components/Channel/Channel'
```
Comment (madsroskar) [inline projects/WhatsAppClone/src/utils/WhatsAppChannelWrapper.tsx:17]: autoimport. Should maybe also be a type import?

File: projects/WhatsAppClone/src/utils/WhatsAppChannelWrapper.tsx [hunk 30]
```diff
@@ -0,0 +1,101 @@
...
+      channel={channel}
+      messageActions={param => {
+        const {
+          pinMessage,
+          unpinMessage,
+          quotedReply,
+          deleteMessage,
+          copyMessage,
+        } = param
+        //  todo: should be toggle pin
```
Comment (madsroskar) [inline projects/WhatsAppClone/src/utils/WhatsAppChannelWrapper.tsx:69]: Maybe move to readme? or remove

**Review by madsroskar — APPROVED** [review]
LGTM, comments are reminders from going through this in a huddle - SDK usage looks fine to me

File: projects/WhatsAppClone/src/client.ts (hunk 7 above)
Comment (vanGalilea) [reply projects/WhatsAppClone/src/client.ts:7]: Our `Chat` component expects a `StreamChat<StreamChatGenerics>` type under and it just does not match with what `getInstance` gives back 
```
StreamChat<
    AttachmentType,
    ChannelType,
    CommandType,
    EventType,
    MessageType,
    ReactionType,
    UserType
  >
```

File: projects/WhatsAppClone/src/components/channel/MessageInput.tsx (hunk 14 above)
Comment (vanGalilea) [reply projects/WhatsAppClone/src/components/channel/MessageInput.tsx:145]: Our `AutoCompleteInput` expects `setInputBoxRef: (ref: TextInput | null) => void;`

Where my local context messageInputRef is `RefObject<TextInput>` originating from a regular `useRef` hook `const messageInputRef = useRef<TextInput>(null)`

Any ideas on how to solve this?

Comment (vanGalilea) [reply projects/WhatsAppClone/src/components/channel/MessageInput.tsx:145]: `useRef` will always wrap `RefObject` 🤔

File: projects/WhatsAppClone/src/App.tsx [hunk 31]
```diff
@@ -0,0 +1,103 @@
...
+          setChannel,
+          selectedChannelsForEditing,
+          setSelectedChannelsForEditing,
+          selectedMessageIdsEditing,
+          setSelectedMessageIdsEditing,
+        }}>
+        <OverlayProvider
+          bottomInset={bottom}
+          value={{style: theme}}
+          // @ts-ignore
```
[the remark this answers is not in the document]
Comment (vanGalilea) [reply projects/WhatsAppClone/src/App.tsx]: This is not typed correctly in the SDK, but I apparently not needing it, I'll create an issue for this

File: projects/WhatsAppClone/src/client.ts (hunk 7 above)
Comment (vanGalilea) [reply projects/WhatsAppClone/src/client.ts:12]: I will create an issue for the SDK

File: projects/WhatsAppClone/src/components/channel/MessageInput.tsx (hunk 14 above)
Comment (vanGalilea) [reply projects/WhatsAppClone/src/components/channel/MessageInput.tsx:145]: This should be wrapped in a RefOvject in our SDK, improve typing- create issue

**Review by madsroskar — APPROVED** [review]
LGTM 👍🏻
